src/BackupExplorer/RestServer/BackupParserManager.cs

```diff
@@ -1,103 +1,103 @@
+
+using Microsoft.ServiceFabric.ReliableCollectionBackup.Parser;
+using System.Runtime.Remoting.Channels;
+
+namespace Microsoft.ServiceFabric.ReliableCollectionBackup.RestServer
+{
+    /// <summary>
+    /// Wrapper around BackupParser for enumerating over transactions.
+    /// </summary>
+    public class BackupParserManager : IDisposable
+    {
+        private static readonly ILog log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
+        public BackupParserManager(BackupParser backupParser)
+        {
+            this.BackupParser = backupParser;
+            //var serializerListed = false;
+
+            // todo : take NumMaxTransactionsInMemory from config.
+            this.BackupParser.TransactionApplied += (sender, args) =>
+            {
+                transactionsQueue = args;
+                Console.WriteLine("Transaction Id {0}, CommitSequenceNumber {1}, Changes {2}", args.TransactionId, args.CommitSequenceNumber, args.Changes.Count());
+            };
+        }
+
+        /// <summary>
+        /// Starts parsing the backup without blocking.
+        /// </summary>
+        /// <returns>Task representing the asynchronous parsing of backup.</returns>
+        public Task StartParsing()
+        {
+            if (this.parsingTask == null)
+            {
+                // todo : take cancellation timeout from config.
+                this.parsingTask = this.BackupParser.ParseAsync(CancellationToken.None);
+            }
+
+            return parsingTask;
+        }
+
+        /// <summary>
+        /// Try to give all transactions.
+        /// </summary>
+        /// <returns>Returns list of transactions</returns>
+        public NotifyTransactionAppliedEventArgs GetTransactions()
+        {
+            lock(queueLock)
+            {
+                return transactionsQueue;
+            }
+        }
+
+        ///// <summary>
+        ///// Checks if it has any unconsumed transactions.
+        ///// </summary>
+        ///// <returns>True if we have more transactions otherwise false.</returns>
+        //public bool HasNextTransaction()
+        //{
+        //    return this.transactionsQueue.Count > 0;
+        //}
+
+        /// <summary>
+        /// Checks if parsing operation has finished.
+        /// If this parsing has finished and <see cref="HasNextTransaction"/> also returns false,
+        /// then we will not have any more transactions in future.
+        /// </summary>
+        /// <returns></returns>
+        public bool HasParsingFinished()
+        {
+            return this.parsingTask.IsCompleted;
+        }
+
+        /// <summary>
+        /// Dispose the Backup manager
+        /// </summary>
+        public void Dispose()
+        {
+            this.BackupParser.Dispose();
+        }
+
+        internal BackupParser BackupParser { get; private set; }
+        
+        private Task parsingTask;
+        
+        private NotifyTransactionAppliedEventArgs transactionsQueue;
```

Comment:
>transactionsQueue [](start = 50, length = 17)

What is transaction queue right now?

Reply:
TransactionsQueue stores  list of all transactions after the checkpoint. Maybe we can rename it to transactionsList

Reply:
So why was it a collection earlier and now only a list in args?

---
In reply to: [484926070](https://github.com/microsoft/service-fabric-backup-explorer/pull/25#discussion_r484926070) [](ancestors = 484926070)

Reply:
Earlier, we used to store dictionarychanged args per transactions. Now after Ashish reported the bug, we are shifting to store all dictionarychanged events in as one, and not as a collection of dictionarychanged args per transaction committed.   We had a choice to either totally remove this, or to change it from a collection  per Transaction committed to  collection of all changes together. I took the 2nd approach.

Reply:
Where are the code changes for that?

---
In reply to: [484975595](https://github.com/microsoft/service-fabric-backup-explorer/pull/25#discussion_r484975595) [](ancestors = 484975595)

Reply:
Not clearing up the changes collected during a transaction, after the next transaction starts did the job. Checkout BackupParserImp for removal of OnTransacationCompleted function removal.